# Проект по SQL 

# Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Компания решила быть на волне и купила крупный сервис для чтения книг по подписке.

# Цель проекта

**Необходимо проанализировать базу данных с информацией о книгах, издательствах, авторах, а также пользовательские обзоры книг и сформулировать ценностное предложение для нового продукта.**

# Задания

- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитать количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;
- Определить автора с самой высокой средней оценкой книг — учитывайть только книги с 50 и более оценками;
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

# Описание данных

**Таблица books**

Содержит данные о книгах:

- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id— идентификатор издателя.

**Таблица authors**

Содержит данные об авторах:

- author_id — идентификатор автора;
- author — имя автора.

**Таблица publishers**

Содержит данные об издательствах:

- publisher_id — идентификатор издательства;
- publisher — название издательства;

**Таблица ratings**

Содержит данные о пользовательских оценках книг:

- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

**Таблица reviews**

Содержит данные о пользовательских обзорах на книги:

- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя пользователя, написавшего обзор;
- text — текст обзора.

## Подключение к базе данных 

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Исследуем таблицы — выведем первые строки

### Таблица books

In [2]:
# чтобы выполнить SQL-запрос, используем Pandas
query = ''' SELECT * 
            FROM books '''

con=engine.connect()

books = pd.io.sql.read_sql(sql=text(query), con = con)

display(books.head())
display(books.info())

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

### Таблица authors

In [3]:
query = ''' SELECT * 
            FROM authors '''

con=engine.connect()

authors = pd.io.sql.read_sql(sql=text(query), con = con)

display(authors.head())
display(authors.info())

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

### Таблица publishers

In [4]:
query = ''' SELECT * 
            FROM publishers '''

con=engine.connect()

publishers = pd.io.sql.read_sql(sql=text(query), con = con)

display(publishers.head())
display(publishers.info())

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

### Таблица ratings

In [5]:
query = ''' SELECT * 
            FROM ratings '''

con=engine.connect()

ratings = pd.io.sql.read_sql(sql=text(query), con = con)

display(ratings.head())
display(ratings.info())

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

### Таблица reviews

In [6]:
query = ''' SELECT * 
            FROM reviews '''

con=engine.connect()

reviews = pd.io.sql.read_sql(sql=text(query), con = con)

display(reviews.head())
display(reviews.info())

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

## Задания

### Посчитаем количество книг, которые вышли после 1 января 2000 года

In [7]:
query = ''' SELECT COUNT (*)
            FROM books 
            WHERE publication_date >= '01-01-2000' ''' 

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

count
0    821

**Вывод:**

821 книг было выпущено после 1 января 2000 года.

### Для каждой книги посчитаем количество обзоров и среднюю оценку

In [19]:
query = ''' SELECT books.book_id, rev.count, rat.mean_rating
            FROM books
            
            LEFT OUTER JOIN (SELECT book_id, COUNT (*) AS count
                             FROM reviews
                             GROUP BY book_id) AS rev 
                             ON books.book_id = rev.book_id 
                             
            LEFT OUTER JOIN (SELECT book_id, AVG (rating) AS mean_rating
                             FROM ratings
                             GROUP BY book_id) AS rat 
                             ON books.book_id = rat.book_id '''

pd.io.sql.read_sql(query, con = engine).sample(10)

book_id  count  mean_rating
370      684    2.0     3.500000
357      865    5.0     3.945946
96       164    2.0     2.500000
416      232    5.0     3.416667
821      253    4.0     3.250000
640      897    2.0     4.000000
707      139    2.0     3.000000
47       644    3.0     3.250000
283      188    3.0     3.000000
625      982    3.0     4.000000

**Вывод:**

По таблице видно, что средний рейтинг книг сильно различается. На первый взгляд, на книги было написано несколько обзоров. 

### Определим, какое издательство выпустило наибольшее число книг толще 50 страниц

In [20]:
query = ''' SELECT pub.publisher, COUNT(*) AS number_of_books
            FROM books
            
            LEFT OUTER JOIN publishers AS pub ON books.publisher_id = pub.publisher_id 
            
            WHERE num_pages >= 50
            GROUP BY pub.publisher
            ORDER BY COUNT(*) DESC '''

pd.io.sql.read_sql(query, con = engine).head(1)

publisher  number_of_books
0  Penguin Books               42

**Вывод:**

Издательство Penguin Books выпустило 42 книги более 50 страниц. Это самый большой показатель среди остальных издательств.

### Определим автора с самой высокой средней оценкой книг, у которых 50 и более оценок

In [21]:
query = ''' SELECT author, AVG (avg) AS mean_rating 
            FROM books
            
            INNER JOIN (SELECT book_id, AVG(rating)
                        FROM ratings
                        GROUP BY book_id
                        HAVING COUNT(book_id) >= 50) AS qwe ON qwe.book_id = books.book_id 
                  
            INNER JOIN (SELECT author_id, author 
                        FROM authors) AS a ON a.author_id = books.author_id 
                  
            GROUP BY author 
            ORDER BY AVG (avg) DESC '''

pd.io.sql.read_sql(query, con = engine).head()

author  mean_rating
0         J.K. Rowling/Mary GrandPré     4.283844
1  Markus Zusak/Cao Xuân Việt Khương     4.264151
2                     J.R.R. Tolkien     4.258446
3                  Louisa May Alcott     4.192308
4                       Rick Riordan     4.080645

**Вывод:**

Самая высокая средняя оценка книг у J.K. Rowling - 4,2.

### Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [ ]:
query = ''' SELECT AVG (count)
            FROM (SELECT username, COUNT (*)
                  FROM reviews
                  WHERE username IN (SELECT username
                                     FROM ratings
                                     GROUP BY username
                                     HAVING COUNT(*) > 48)
                  GROUP BY username) AS qwe
            
        '''

pd.io.sql.read_sql(query, con = engine)

**Вывод:**

Среднее количество обзоров пользователей, которые поставили больше 48 оценок - 24.

## Вывод:

Была проанализирована база данных с информацией о книгах, издательствах, авторах, а также пользовательские обзоры книг.

**В результате анализа базы данных было выявлено:**

- 821 книг вышли после 1 января 2000 года 

**В данном сервисе присутствует большое количество свежих книг, которые были выпущены за последние 20 лет**

- Для каждой книги посчитали количество обзоров и среднюю оценку. 
- Среднее количество обзоров пользователей, которые поставили больше 48 оценок - 24 обзора.

**Пользователи сервиса активно принимают участие в обзорах и оценках книг.**

- Издательство, которое выпустило наибольшее число книг толще 50 страниц - Penguin Books. Оно опубликовало 42 книги более 50 страниц. 
- Автор с самой высокой оценкой книг - J.K. Rowling. Средняя оценка - 4,2.

**Также в данном сервисе присутствуют книги популярных авторов и издательств.**